scripts/mimc_constraints.sage

```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/sage
+
+"""
+Computes the valid configurations of mimc for ALT-BN128 and BLS12-377 and
+the number of multiplications required.
+"""
+
+def mults_to_combine_num_powers(num_powers):
+    num_powers_bits = num_powers.bits()
+    num_powers_true_bits = sum(num_powers_bits)
+    assert 0 != num_powers_true_bits
+
+    # Compute largest pow2 group of exponents
+    pow2_group = 1 << (len(num_powers_bits) - 1)
+    # remaining exponents
+    num_powers = num_powers - pow2_group
+    # cost to combine the pow2_group into a single power
+    mults = pow2_group - 1
+
+    # No more powers to multiply together
+    if num_powers == 0:
+        return mults
+
+    # Total cost:
+    #   mults + <cost to combine this with remaining powers>
+    return mults + mults_to_combine_num_powers(1 + num_powers)
+
+
+def compute_mults(e):
+    """
+    For e = 2^t - 1, mults can be computed as 2*t - 2, but we implement a more
+    general function.
+    """
+    e_bits = e.bits()
+    e_true_bits = sum(e_bits)   # num true bits
+    # 1 multiplication for each exponent used
+    mults = len(e_bits) - 1
+    # combine e_true_bits
+    return mults + mults_to_combine_num_powers(e_true_bits)
+
+
+def output_valid_config_and_constraints(r, log_2_r, e):
+    if 1 == gcd(r-1, e):
+        mults = compute_mults(e)
+        rounds = ceil(log_2_r / log(e, 2))
+        constraints = mults * rounds + 1
+        print(f"  e={e}, rounds={rounds}, mults={mults}, contraints={constraints}")
```

Comment:
I assume that your version of sage is >9.0, right?

Reply:
Yes.  I installed with `brew`, which only provides 9.2 at the current time.

Reply:
Makes sense, I had an older sage installation (`<0.9`) which wasn't supporting python3 yet.